core/types/src/quote.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Quote types
+
+use crate::{
+    attestation_key::QuoteSignatureType, impl_newtype_for_bytestruct, new_type_accessors_impls,
+    IsvSvn, ReportBody,
+};
+use core::mem;
+use mc_sgx_core_sys_types::{
+    sgx_basename_t, sgx_epid_group_id_t, sgx_quote_nonce_t, sgx_quote_sign_type_t,
+    sgx_report_body_t,
+};
+
+/// EPID Group ID
+#[derive(Default, Debug, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct EpidGroupId(sgx_epid_group_id_t);
+
+new_type_accessors_impls! {
+    EpidGroupId, sgx_epid_group_id_t;
+}
+
+const BASENAME_SIZE: usize = 32;
+
+/// Basename
+#[derive(Default, Debug, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Basename(sgx_basename_t);
+
+impl_newtype_for_bytestruct! {
+    Basename, sgx_basename_t, BASENAME_SIZE, name;
+}
+
+const NONCE_SIZE: usize = 16;
+
+/// Quote Nonce
+#[derive(Default, Debug, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct QuoteNonce(sgx_quote_nonce_t);
+
+impl_newtype_for_bytestruct! {
+    QuoteNonce, sgx_quote_nonce_t, NONCE_SIZE, rand;
+}
+
+/// The raw bytes representing a quote.
+///
+/// Should not be used directly instead use [`Quote`].
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct RawQuote<'a> {
+    bytes: &'a [u8],
+}
+
+#[derive(Default, Debug, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Version(u16);
+
+new_type_accessors_impls! {
+    Version, u16;
+}
+
+pub trait BaseQuote {
+    /// Provides access to the [`RawQuote`] to perform the common lookup
+    /// operations on the basic quote type.
+    fn raw_quote(&self) -> &RawQuote;
+
+    /// Version of the quote
+    fn version(&self) -> Version {
+        let bytes = self.raw_quote().bytes[..2]
+            .try_into()
+            .expect("Quote bytes aren't big enough to hold `version`");
+        u16::from_le_bytes(bytes).into()
+    }
+
+    /// The signature type
+    fn signature_type(&self) -> QuoteSignatureType {
+        let bytes = self.raw_quote().bytes[2..4]
```

Comment:
I'm not sure what thoughts are on the explicit offsets right here.
Since these are only used here, I think declaring const for each offset would be worse:
- It disassociates the value and the usage too much.  I.e. if someone wants to know what the value is they need to look elsewhere.
- it creates it's own form of code bloat, big lists of constants that people have to navigate past to get to the code of interest.

ideally these would something like `[offset_of(field)..size_of(field)]` but rust lacks an offset of.
There is https://docs.rs/memoffset/latest/memoffset/macro.span_of.html if we want to pull in a dep for this